torch/lib/c10d/test/ProcessGroupGlooTest.cpp

```diff
@@ -252,6 +252,24 @@ void testBroadcast(const std::string& path, const at::Backend b) {
   }
 }
 
+void testBarrier(const std::string& path) {
+  const auto size = 2;
+  auto tests = CollectiveTest::initialize(path, size);
+
+  // Kick off work
+  std::vector<std::shared_ptr<::c10d::ProcessGroup::Work>> work(size);
+  for (auto i = 0; i < size; i++) {
+    work[i] = tests[i].getProcessGroup().barrier();
+  }
+
+  // Wait for work to complete
+  for (auto i = 0; i < size; i++) {
+    if (!work[i]->wait()) {
+      throw work[i]->exception();
+    }
+  }
+}
```

Comment:
This doesn't really test that the barrier works, but I guess it's ok as a smoke test

Reply:
Yep it only tests that the two things returned from barrier complete. To test whether it's a real barrier we would need to induce race conditions and explicitly test that adding a barrier removes those races -- not worth the squeeze IMO.